tcl8.3.4/doc/GetCwd.3
'\"
'\" Copyright (c) 1998-1999 Scriptics Corportation
'\"
'\" See the file "license.terms" for information on usage and redistribution
'\" of this file, and for a DISCLAIMER OF ALL WARRANTIES.
'\" 
'\" RCS: @(#) $Id: GetCwd.3,v 1.1 2001/11/05 20:06:32 jgdavidson Exp $
'\" 
.so man.macros
.TH Tcl_GetCwd 3 8.1 Tcl "Tcl Library Procedures"
.BS
.SH NAME
Tcl_GetCwd, Tcl_Chdir \- manipulate the current working directory
.SH SYNOPSIS
.nf
\fB#include <tcl.h>\fR
.sp
char *
\fBTcl_GetCwd\fR(\fIinterp\fR, \fIbufferPtr\fR)
.sp
int
\fBTcl_Chdir\fR(\fIpath\fR)
.SH ARGUMENTS
.AS Tcl_DString *bufferPtr
.AP Tcl_Interp *interp in
Interpreter in which to report an error, if any.
.AP Tcl_DString *bufferPtr in/out
This dynamic string is used to store the current working directory.
At the time of the call it should be uninitialized or free.  The
caller must eventually call \fBTcl_DStringFree\fR to free up
anything stored here.
.AP char *path in
File path in UTF\-8 format.
.BE

.SH DESCRIPTION
.PP
These procedures may be used to manipulate the current working
directory for the application.  They provide C\-level access to
the same functionality as the Tcl \fBpwd\fR command.
.PP
\fBTcl_GetCwd\fR returns a pointer to a string specifying the current
directory, or NULL if the current directory could not be determined.
If NULL is returned, an error message is left in the interp's result.
Storage for the result string is allocated in bufferPtr; the caller
must call \fBTcl_DStringFree()\fR when the result is no longer needed.
The format of the path is UTF\-8.
.PP
\fBTcl_Chdir\fR changes the applications current working directory to
the value specified in \fIpath\fR.  The format of the passed in string
must be UTF\-8.  The function returns -1 on error or 0 on success.

.SH KEYWORDS
pwd


ChangeLog
2004-06-14 20:28  Dossy Shiobara <dossy@panoptic.com>

	* nsd/connio.c: If nsend == 0, we would never call Ns_WriteConn to
	  flush the queued headers.  This could happen when we're sending
	  zero bytes of data as a response.  This fixes bug #971016 in the
	  case where fastpath.cache=false and fastpath.mmap=false and a
	  zero byte file is requested.

2004-06-14 20:25  Dossy Shiobara <dossy@panoptic.com>

	* nsd/return.c: unnecessary test for data != NULL actually caused
	  part of bug #971016 when fastpath.cache=false and
	  fastpath.mmap=true and the file requested is zero bytes, mmap()
	  returns 0 which gets passed along as data == NULL, causing
	  ReturnCharData() to not flush the queued headers.  removing the
	  if is safe as Ns_WriteConn will simply flush any queued data

2004-06-08  Rob Crittenden <rcrittenden0569@aol.com>

	* Makefile: remove Makefile.module on distclean.
	* include/ns.h, nsd/log.c: Allow users to override logging functions.
	* nsd/str.c: Fix crash bug in Ns_Trim* when trimming NULL strings.

2004-06-08  Rob Crittenden <rcrittenden0569@aol.com>

	* TAG aolserver_v40_r5

2004-03-09 13:35  Rob Crittenden <rcrittenden0569@aol.com>

	* nsd/tclhttp.c: The last header was being missed because the pointer
	  was being advanced past the \r\n.

2004-02-26 14:37  Dossy Shiobara <dossy@panoptic.com>

	* nsd/unix.c: uid_t is supposed to be unsigned, but apparently
	  isn't on OSX, apparently.

2004-02-25 19:56  Dossy Shiobara <dossy@panoptic.com>

	* nscgi/nscgi.c: Ns_ConnSetRequiredHeaders now requires length < 0
	  if you don't want a Content-Length header

2004-02-25 19:55  Dossy Shiobara <dossy@panoptic.com>

	* nsd/return.c: fix so that it's still possible to formulate a
	  response with no Content-Length header (such as for nscgi, or
	  HTTP streaming) where you don't know the Content-Length header in
	  advance

2004-02-24 07:15  Dossy Shiobara <dossy@panoptic.com>

	* nsd/: nsmain.c, nsd.h, driver.c: fix race condition where
	  DriverThreads are spawned but before they can Ns_SockListenEx,
	  the main thread does its NsClosePreBound and closes the prebound
	  sockets, making them unavailable to the DriverThreads if they
	  require root priv to listen on

2004-02-18 01:30  Dossy Shiobara <dossy@panoptic.com>

	* nsperm/nsperm.c: fixed bug #899364, restoring old nsperm
	  functionality as it was in 3.x.

2004-02-15  Jim Davidson <jgdavidson@aol.com>

	* nsd/conn.c:
	* include/ns.h:
	* nsd/return.c: Added new Ns_ConnGet/SetKeepAliveFlag APIs to
	allow custom request procedures to enable keep-alive directly.

	* nsd/driver.c: Minor updates to separate conn & driver structs.

2004-01-01  Scott S. Goodwin <scott@scottg.net>

        * nsd/adpparse.c: fixed bug #868600, where tag attribute values
	encapsulated within single quotes weren't being parsed properly.

2003-12-27  Scott S. Goodwin <scott@scottg.net>

        * nsd/urlopen.c: fixed bug #844809.

        * sample-config.tcl: fixed bug #805761.

2003-12-24  Scott S. Goodwin <scott@scottg.net>

	* include/nsthread.h: Bumped up NS_THREAD_NAMESIZE from 32 to 64.
	Mutex names were being truncated, making it impossible to distinguish
	mutexes with longer names when looking at stats.

        * nscgi/nscgi.c: Changed HTTPS env variable to be set to "on" or "off"
	instead of "1" or "0". There doesn't appear to be any standard for
	this, but Apache uses the on/off string in the environment.

2003-12-16  Scott S. Goodwin <scott@scottg.net>

        * nscgi/nscgi.c: Fixed issue with HTTPS environment variable not being
        set, and port not being set to 443 if connection is SSL'd. Manually
        applied these changes from a patch supplied by Vinod Kurup.

2003-11-16  Jim Davidson <jgdavidson@aol.com>

	* include/ns.h: Update version ids, added NS_SOCK_DROP and
	NS_TCL_TRACE constants and pre-queue function definitions.

	* nsd/nsd.h:  Added new "times" struct to Conn for more
	fine-grained timing.  Need to provide options to enable
	these time stamps plus ways to get the data.

	* nsd/driver.c:  Restructured in a few ways:

	1. Each loaded driver now has it's own thread.  This should
	allow some more concurrency (if necessary) and generally
	made the code more manageable.

	2. New pre-queue I/O callback API.  The API will be documented
	with an example HTTP module later.

	3. Conn allocation is handled by the driver threads and
	request buffers have been moved to the Conn from the Sock.

	4. Reads for synchrounous drivers (i.e., SSL) are handled
	by a pool of reader threads instead of by the conn threads.
	The number of threads is set with a "maxreaders" config
	option.  This change should enable both Host: header based
	virtual servers and pre-queue callbacks for SSL connections.

	5. New "maxsocks" config for drivers is now used to manage
	the concurrency limit of the server.  Conn's are now queued
	without limit into the connection pools and the previous
	"maxconns" virtual-server config is ignored.

	* nsd/queue.c: Cleaned up "connsperthread" code and updated
	NsQueueConn to support new driver code.

	* nsd/conn.c:
	* nsd/connio.c:
	* nsd/form.c:
	* nsd/nsd.h:
	* nsd/queue.c:
	* nsd/return.c: Moved connection request buffers from Sock
	to Conn structure.

	* nsd/server.c: Removed startup Conn pre-allocation as Conns
	are now allocated dynamically by the driver.  Also, removed
	the "maxconns" parameter which is no longer used.

	* nssock/nssock.c: Added new "async" option (default on)
	mostly to test new reader thread capabilities of the core
	driver code.

	* nsd/tclinit.c: Added a new generalized Tcl interp tracing
	facility and cleaned up some previous init code.  This
	should make it easier to support standard Tcl extensions
	and will be documented with an example module later.

	* nsd/nsmain.c: Fixed order of driver shutdown wait.

	* nsd/init.c: Fixed order of init calls.

	* nsd/binder.c:
	* nsd/dns.c: Named unnamed mutexes.

	* nsd/tclset.c:
	* nsd/tclsock.c: Removed unused functions.


tcl7.6/compat/getcwd.c
/* 
 * getcwd.c --
 *
 *	This file provides an implementation of the getcwd procedure
 *	that uses getwd, for systems with getwd but without getcwd.
 *
 * Copyright (c) 1993 The Regents of the University of California.
 * Copyright (c) 1994 Sun Microsystems, Inc.
 *
 * See the file "license.terms" for information on usage and redistribution
 * of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 *
 * SCCS: @(#) getcwd.c 1.5 96/02/15 12:08:20
 */

static const char *RCSID = "@(#) $Header: /Users/dossy/Desktop/cvs/aolserver/tcl7.6/compat/Attic/getcwd.c,v 1.1.1.1 2000/05/02 13:48:25 kriston Exp $, compiled: " __DATE__ " " __TIME__;

#include "tclInt.h"
#include "tclPort.h"

extern char *getwd _ANSI_ARGS_((char *pathname));

char *
getcwd(buf, size)
    char *buf;			/* Where to put path for current directory. */
    size_t size;		/* Number of bytes at buf. */
{
    char realBuffer[MAXPATHLEN+1];
    int length;

    if (getwd(realBuffer) == NULL) {
	/*
	 * There's not much we can do besides guess at an errno to
	 * use for the result (the error message in realBuffer isn't
	 * much use...).
	 */

	Tcl_SetErrno (EACCES);
	return NULL;
    }
    length = strlen(realBuffer);
    if (length >= size) {
	Tcl_SetErrno (ERANGE);
	return NULL;
    }
    strcpy(buf, realBuffer);
    return buf;
}

